### PR TITLE
Issue #118 - app delete should be done through controller using finalizers

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/118

### CODE REVIEW

**Review by merenbach — APPROVED** [review]
:+1:

File: server/application/application.proto [hunk 1]
```diff
@@ -22,9 +22,7 @@ message ApplicationResponse {}
 
 message DeleteApplicationRequest {
 	string name = 1;
-	string namespace = 2;
-	string server = 3;
-	bool force = 4;
+	bool force = 2;
```
Comment (jessesuen) [inline server/application/application.proto]: The term force is unclear what will happen in the delete request. This flag really indicates whether or not to perform cascaded deletion or not.

Can we rename this to `cascade` which can either be nil, true or false?

I think the semantics should be:
* If cascade is false, remove the finalizer (if present). delete the object directly from kubernetes. this should not involve the controller.
* If cascade is true, set the finalizer (if not already set). delete the object directly from kubernetes. deletion will be blocked until the finalizer is removed by the controller
* If cascade is nil. don't touch finalizers one way or the other. delete the object directly from kubernetes. deletion may or may not involve controller depending on if finalizer was already set.

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -285,6 +286,33 @@ type RepositoryList struct {
...
+	for i, finalizer := range app.Finalizers {
+		if finalizer == name {
+			return i
+		}
+	}
+	return -1
+}
+
+// NeedPruneResources indicates if resources finalizer is set and controller should delete app resources before deleting app
+func (app *Application) NeedPruneResources() bool {
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: I think prune may be incorrect terminology.  Pruning implies something will be left over (e.g. your have resources that is no longer managed in git and you want to delete those). But this field indicates whether or not you should delete all the dependant resources.

How about:
app.CascadeDelete()
app.OrphanDependants()

File: pkg/apis/application/v1alpha1/types.go [hunk 3]
```diff
@@ -285,6 +286,33 @@ type RepositoryList struct {
...
+	return app.getFinalizerIndex(common.ResourcesFinalizerName) > -1
+}
+
+// SetNeedPruneResources sets or remove resources finalizer
+func (app *Application) SetNeedPruneResources(prune bool) {
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go]: I think SetCascadedDeletion(true) is better.

File: server/application/application.proto (hunk 1 above)
Comment (jessesuen) [reply server/application/application.proto]: I think to support nullable fields you may need (gogoproto.nullable) = true.

File: controller/controller.go [hunk 4]
```diff
@@ -202,31 +203,82 @@ func retryUntilSucceed(action func() error, desc string, ctx context.Context, ti
...
+				_, err = ctrl.applicationClientset.ArgoprojV1alpha1().Applications(app.Namespace).Patch(app.Name, types.MergePatchType, patch)
+			}
+		}
 	}
-	if app.Operation == nil {
-		return true
+	if err != nil {
+		log.Errorf("Unable to delete application resources: %v", err)
+	} else {
+		log.Infof("Successfully deleted resources for application %s", app.Name)
```
Comment (jessesuen) [inline controller/controller.go]: If we fail to delete the app dependencies, the user has no visibility of why the app won't be deleted in the event of an error. Should we update HealthStatus here?

**Review by jessesuen — CHANGES_REQUESTED** [review]

**Comment by alexmt** [comment]
@jessesuen your comments are addressed. PTAL

File: server/application/application.go [hunk 5]
```diff
@@ -102,47 +102,46 @@ func (s *Server) Update(ctx context.Context, a *appv1.Application) (*appv1.Appli
...
-		if err != nil && !q.Force {
-			return nil, err
-		}
-		if clst != nil {
-			config := clst.RESTConfig()
-			err = kube.DeleteResourceWithLabel(config, namespace, common.LabelApplicationName, q.Name)
-			if err != nil && !q.Force {
+	if q.Cascade != nil {
+		if *q.Cascade != a.CascadedDeletion() {
+			a.SetCascadedDeletion(true)
```
Comment (jessesuen) [inline server/application/application.go]: This seems like a bug? If q.Cascade is false, but a.CascadedDeletion is true,  then we always set cascaded deletion to be true and do not honor the request from the API.

**Review by jessesuen — CHANGES_REQUESTED** [review]

**Review by jessesuen — APPROVED** [review]

**Pull request merged**